shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm

```diff
@@ -355,6 +355,61 @@ static FlutterAutofillType autofillTypeOf(NSDictionary* configuration) {
   return FlutterAutofillTypeNone;
 }
 
+static BOOL isApproximatelyEqual(float x, float y, float delta) {
+  return x >= y - delta && x <= y + delta;
+}
+
+// Checks whether point should be considered closer to selectionRect compared to
+// otherSelectionRect.
+//
+// If checkRightBoundary is set, the right-center point on selectionRect and
+// otherSelectionRect will be used instead of the left-center point.
+//
+// This uses special (empirically determined) logic for determining the closer rect,
+// rather than a simple distance calculation. First, the closer vertical distance is
+// determined. Within the closest y distance, if the point is above the bottom
+// of the closest rect, the x distance will be minimized; however, if the point is
+// below the bottom of the rect, the x value will be maximized.
+static BOOL isSelectionRectCloserToPoint(CGPoint point,
+                                         CGRect selectionRect,
+                                         CGRect otherSelectionRect,
+                                         BOOL checkRightBoundary) {
+  CGPoint pointForSelectionRect =
+      CGPointMake(selectionRect.origin.x + (checkRightBoundary ? selectionRect.size.width : 0),
+                  selectionRect.origin.y + selectionRect.size.height * 0.5);
+  float yDist = fabs(pointForSelectionRect.y - point.y);
+  float xDist = fabs(pointForSelectionRect.x - point.x);
+
+  CGPoint pointForOtherSelectionRect =
+      CGPointMake(otherSelectionRect.origin.x + (checkRightBoundary ? selectionRect.size.width : 0),
+                  otherSelectionRect.origin.y + otherSelectionRect.size.height * 0.5);
+  float yDistOther = fabs(pointForOtherSelectionRect.y - point.y);
+  float xDistOther = fabs(pointForOtherSelectionRect.x - point.x);
+
+  BOOL isCloserVertically = yDist < yDistOther - 1;
+  BOOL isEqualVertically = isApproximatelyEqual(yDist, yDistOther, 1);
+  BOOL isAboveBottomOfLine = point.y <= selectionRect.origin.y + selectionRect.size.height;
+  BOOL isCloserHorizontally = xDist <= xDistOther;
+  BOOL isBelowBottomOfLine = point.y > selectionRect.origin.y + selectionRect.size.height;
+  BOOL isFartherToRight =
+      selectionRect.origin.x + (checkRightBoundary ? selectionRect.size.width : 0) >
+      otherSelectionRect.origin.x;
+  return (isCloserVertically ||
+          (isEqualVertically && ((isAboveBottomOfLine && isCloserHorizontally) ||
+                                 (isBelowBottomOfLine && isFartherToRight))));
+}
+
+// Checks whether Scribble features are possibly available – meaning this is an iPad running iOS 14+
+static BOOL isScribbleAvailable() {
+  if (@available(iOS 14.0, *)) {
+    NSString* deviceModel = (NSString*)[UIDevice currentDevice].model;
+    if ([[deviceModel substringWithRange:NSMakeRange(0, 4)] isEqualToString:@"iPad"]) {
```

Comment:
I was wondering if we had a better way to do this, but I couldn't find one.  I didn't see anywhere else in the engine where we try to detect iPad.  I guess this is good enough.

Reply:
`if ( UI_USER_INTERFACE_IDIOM() == UIUserInterfaceIdiomPad )` is another way